src/app/auth/auth.interceptor.ts

```diff
@@ -0,0 +1,30 @@
+import { Injectable } from '@angular/core';
+import { HttpRequest, HttpEvent, HttpInterceptor, HttpHandler, HttpErrorResponse } from '@angular/common/http';
+import { catchError, Observable, of, throwError } from 'rxjs';
+import { Router } from '@angular/router';
+import { MatSnackBar } from '@angular/material/snack-bar';
+import { showSnackbar } from '../shared/snackbar-utils';
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+
+    constructor(
+        private router: Router,
+        private snackBar: MatSnackBar
+    ) { }
+
+    private handleAuthError(err: HttpErrorResponse): Observable<any> {
+        //handle your auth error or rethrow
+        if (err.status === 401 || err.status === 403) {
+            showSnackbar(this.snackBar, "Your session has expired, please log in again")
+            // Redirect to the login page
+            this.router.navigateByUrl(`/login`);
+            return of(err.message);
+        }
+        return throwError(() => err);
+    }
+
+    intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+        return next.handle(req).pipe(catchError(x => this.handleAuthError(x)));
```

Comment:
你可以在這裡把每個 request 攔截下來加上 credential。就不用每個 api 都要自己去加入
```suggestion
        req = req.clone({
          withCredentials: true
        });
        return next.handle(req).pipe(catchError(x => this.handleAuthError(x)));
```